library/core/src/num/mod.rs

```diff
@@ -846,37 +849,79 @@ fn from_str_radix<T: FromStrRadixHelper>(src: &str, radix: u32) -> Result<T, Par
     };
 
     let mut result = T::from_u32(0);
-    if is_positive {
-        // The number is positive
-        for &c in digits {
-            let x = match (c as char).to_digit(radix) {
-                Some(x) => x,
-                None => return Err(PIE { kind: InvalidDigit }),
-            };
-            result = match result.checked_mul(radix) {
-                Some(result) => result,
-                None => return Err(PIE { kind: PosOverflow }),
-            };
-            result = match result.checked_add(x) {
-                Some(result) => result,
-                None => return Err(PIE { kind: PosOverflow }),
-            };
+    if mem::size_of::<T>() <= mem::size_of::<u32>() && radix == 10 {
+        // The ALU can reorder these adds and do more in parallel
+        // as each mul isn't dependent on the previous answer.
+        let factors = &MULTIPLIER[MULTIPLIER.len() - digits.len()..];
+        let mut idx = 0;
+        if is_positive {
+            // The number is positive
+            for &c in digits {
+                let x = match (c as char).to_digit(radix) {
+                    Some(x) => x,
+                    None => return Err(PIE { kind: InvalidDigit }),
+                };
+                let x = match factors[idx].checked_mul(x) {
+                    Some(result) => result,
+                    None => return Err(PIE { kind: PosOverflow }),
+                };
+                result = match result.checked_add(x) {
+                    Some(result) => result,
+                    None => return Err(PIE { kind: PosOverflow }),
+                };
+                idx += 1;
```

Comment:
No need to manage an index manually if you loop over the factors directly:
```rust
for (&c, &factor) in digits.zip(factors) { ... }
```

Assuming zip doesn't inhibit any optimizations.

Reply:
Both `zip` and `enumerate` inhibited optimisations - I like that euphamism.

Reply:
I've seen some weird interactions between `zip` and `rev`, which currently(?) cause suboptimal assembly, but never with `enumerate`. Is this another case of missing `-Copt-level=3` on godbolt?

PS: No euphemism intended.